[숫자 야구 게임] 김수환 pr입니다.

genius00hwan opened the pull request:
예외테스트, 게임종료 후 재 시작 테스트 실패 했습니다.
코드 수정후 다시 재출 하겠습니다.

genius00hwan committed:
protorype

genius00hwan committed:
Target 클래스 완성

genius00hwan committed:
feat(Application): add function to make target number

genius00hwan committed:
feat(Application): add functions to judge user's score

genius00hwan committed:
feat(Application): add functions play Game

sunwootest commented on src/main/java/baseball/Game.java:
```diff
@@ -0,0 +1,70 @@
+public class Game {
+
+	Vector<Integer> in = new Vector<Integer>();
+	Judge judge = new Judge();
```
자바 빈 규약에 따른 클래스의 구성요소를 떠올리면 이렇게 필드를 선언하고 초기화 하는 것은 부자연스럽습니다. 물론 자바 빈 규약을 완전히 지킬 필요는 없지만 이런 방식은 알고리즘 문제 풀 때의 방식 같습니다.

sunwootest commented on src/main/java/baseball/Game.java:
```diff
@@ -0,0 +1,70 @@
+			judge.scoring(in);
+			printScore(judge);
+		} while (!judge.correct);
+	}
```
do-while문은 지양합시다. 코드 가독성이 떨어집니다.

sunwootest commented on src/main/java/baseball/Game.java:
```diff
@@ -0,0 +1,70 @@
+	Vector<Integer> in = new Vector<Integer>();
+	Judge judge = new Judge();
+
+	public void execution() {
+		int playGame = 1;
+		while (playGame == 1) {
+			this.play();
+			System.out.println("게임 종료");
+			playGame= Integer.parseInt(Console.readLine());
+			if (playGame != 1 && playGame != 2) {
+				throw new IllegalArgumentException();
+			}
+		}
+	}
```
`출력`과 `입력에 대한 검증`을 다른 메소드로 분리해볼까요?

sunwootest commented on src/main/java/baseball/Judge.java:
```diff
@@ -0,0 +1,40 @@
+
+public class Judge {
+
+	Target target = new Target();
```
초기화는 생성자를 통해서 진행하는 것이 낫습니다.

sunwootest commented on src/main/java/baseball/Judge.java:
```diff
@@ -0,0 +1,40 @@
+	int ball;
+	boolean correct;
+
+	public void scoring(Vector<Integer> in) {
+		this.strike = countStrike(in);
+		this.ball = countBall(in) - this.strike;
+		this.correct = (this.strike == numOfDigit);
+	}
```
Java에서 Vector 사용은 지양하는 편입니다. 그 이유에 대해서 알아보시고 다른 대안을 적용해보세요.

genius00hwan replied:
네 수정하도록 하겠습니다.

sunwootest reviewed (changes requested):
알고리즘 문제 푸는 습관이 이번 과제에 그대로 반영된 것 같습니다. 객체지향적으로 설계를 해봅시다. 

요구사항을 충족하는 것이 1순위입니다. 하나의 메소드에 모든 요구사항을 다 포함시키고 리팩토링 하는 방법으로 진행해도 괜찮습니다.

genius00hwan committed:
fix(Application): fix functions play Game
feat(Application): add class of CONSTANTS

genius00hwan committed:
fix(Application): fix functions play Game
fix(Application): fix class of CONSTANTS
feat(Application): add ExceptionHandler class

genius00hwan committed:
fix(Application): fix input function
feat(Application): add
function to handle duplication input

genius00hwan committed:
fix(Application):print Score function move to Judge Class

genius00hwan committed:
fix(Application): construct target on Game class' play() func
feat(Readme) : write README.md file

genius00hwan reviewed (commented):
Readme 파일을 작성하고, 메시지, 상수 값등을 저장하는 클래스를 만들었습니다. 예외처리 클래스도 만들었습니다.
감사합니다

genius00hwan reviewed (commented):
코드 리뷰 해주시면 감사하겠습니다!!

genius00hwan committed:
fix(Application): create getter method of Constants

genius00hwan reviewed (commented):
캡슐화를 위해 getter를 추가했습니다.

genius00hwan committed:
fix(Application): create getter methods for encapsulation

sunwootest commented on src/main/java/baseball/Constants.java:
```diff
@@ -0,0 +1,79 @@
+package baseball;
+
+public class Constants {
+	//MESSAGES
+	private static final String GUESS_NUMBER_MESSAGE = "숫자를 입력해주세요 : ";
+	private static final String END_GAME_MESSAGE = "게임 종료";
+	private static final String BALL_STRING = "볼";
+	private static final String STRIKE_STRING = "스트라이크";
+	private static final String NOTHING_STRING = "낫싱";
+	private static final String SPACE = " ";
+
+	//VALUES
+	private static final int NUM_OF_DIGIT = 3;
+	private static final int MIN_OF_RANGE = 1;
+	private static final char MIN_OF_RANGE_CHAR = '1';
+	private static final int MAX_OF_RANGE = 9;
+	private static final char MAX_OF_RANGE_CHAR = '9';
+
+	private static final char ZERO_CHAR = '0';
+
+	private static final int PLAY_GAME = 1;
+	private static final int END_GAME = 2;
+
```
패키지를 하나 생성해서 여러 클래스로 분리하는 편이 나아보입니다.

sunwootest commented on src/main/java/baseball/ExceptionHandler.java:
```diff
@@ -0,0 +1,32 @@
+
+import java.util.ArrayList;
+
+public class ExceptionHandler {
+
+	public void handleInvalidSelection(Integer selection) {
+		if (!selection.equals(getPlayGame()) && !selection.equals(getEndGame())) {
+			throw new IllegalArgumentException();
+		}
+	}
+
+	public void handleInvalidLength(int length) {
+		if (length != getNumOfDigit()) {
+			throw new IllegalArgumentException();
+		}
+	}
+
+	public void handleOutOfRange(char number) {
+		if (number < getMinOfRangeChar() || number > getMaxOfRangeChar()) {
+			throw new IllegalArgumentException();
+		}
+	}
+
+	public void handleDuplication(char cur, ArrayList<Integer> oldNums) {
+		if (oldNums.contains(cur - getZeroChar())) {
+			throw new IllegalArgumentException();
+		}
+	}
```
예외처리를 별도의 클래스에서 처리해줄 필요가 있을까요?

sunwootest commented on src/main/java/baseball/Game.java:
```diff
@@ -0,0 +1,58 @@
+	public void execution() {
+		playGame = getPlayGame();
+		while (playGame == getPlayGame()) {
+			this.play();
```
`play()`라고만 작성해도 충분합니다.

sunwootest commented on src/main/java/baseball/Game.java:
```diff
@@ -0,0 +1,58 @@
+	public void input() {
+		in = new ArrayList<>();
+		String inStr = "";
+		System.out.print(getGuessNumberMessage());
```
출력을 별도의 메소드로 분리해봅시다. 출력내용을 분리하는 것도 좋지만 sout 전체를 분리하는 편이 나아보입니다.

sunwootest reviewed (changes requested):
오늘 설명해준 내용 바탕으로 객체지향적으로 리팩토링 해보세요! 
리뷰할 수 있는 부분들이 있는데 크게 건드리지 않았습니다 오늘 설명을 듣고나서 개선할 수 있을 것 같아서요.

genius00hwan committed:
feat(Application): create printer class
feat(Application): create Constants Package for Messages and Constants Values

genius00hwan reviewed (commented):
상수 패키지를 따로 만들고 printer 클래스를 따로 만들었습니다.
